File: lib_mvp/src/main/java/com/kasiengao/mvp/java/BasePresenter.java
package com.kasiengao.mvp.java;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;

import java.lang.ref.WeakReference;

/**
 * @ClassName: BasePresenter
 * @Author: KaiSenGao
 * @CreateDate: 2020/3/26 16:01
 * @Description:
 */
public class BasePresenter<V extends BaseContract.BaseView> implements BaseContract.BasePresenter<V> {

    private V mView;

    private WeakReference<V> mWeakReference;

    /**
     * 给子类使用的获取View的操作
     * 不允许复写
     *
     * @return View
     */
    public final V getView() {
        return this.mView;
    }

    @Override
    public void attachView(V rootView) {
        this.mView = rootView;
        this.mWeakReference = new WeakReference<>(rootView);
    }

    @Override
    public void detachView() {
        this.mView = null;
        this.mWeakReference.clear();
    }

    @NonNull
    protected Context getContext() {
        return this.mView.getContext();
    }

    @NonNull
    protected LifecycleOwner getLifecycleOwner() {
        return this.mView.getLifecycleOwner();
    }
}

File: lib_mvp/src/main/java/com/kasiengao/mvp/kt/BasePresenter.kt
package com.kasiengao.mvp.kt

import android.content.Context
import androidx.lifecycle.LifecycleOwner
import java.lang.ref.WeakReference

/**
 * @ClassName: BasePresenter
 * @Author: KaiSenGao
 * @CreateDate: 2020/3/26 17:00
 * @Description:
 */
public open class BasePresenter<V : BaseContract.BaseView> : BaseContract.BasePresenter<V> {

    private var mView: V? = null

    private var mWeakReference: WeakReference<V>? = null

    /**
     * 给子类使用的获取View的操作
     * 不允许复写
     *
     * @return View
     */
    public fun getView(): V = this.mView!!

    override fun attachView(rootView: V) {
        this.mView = rootView
        this.mWeakReference = WeakReference(rootView)
    }

    override fun detachView() {
        this.mView = null
        this.mWeakReference!!.clear()
    }

    protected fun getContext(): Context? = this.mView!!.context

    protected fun getLifecycleOwner(): LifecycleOwner? = this.mView!!.lifecycleOwner
}